### DISCUSSION TITLE
Allow distance constraints between parallel lines

### DISCUSSION THREAD

**Discussion by calebmeyer** [hlorus/CAD_Sketcher#211] [Feature Requests]
# Motivation / Existing Problem
Today, if you have two parallel lines that form say a parallelogram instead of a rectangle, it's hard to get the distance between them 

# Description
I want to be able to select two lines which already have a parallel constraint and set a distance constraint on them.

# Alternative Solutions
To accomplish this today requires a lot of extra steps and constraints:
- you need a line perpendicular to the two lines that are parallel
- you need a point coincident to each of the parallel lines (or their extensions if they're far apart) that is also coincident to the perpendicular line
- those two coincident points can then be measured for distance

# Mockup
If you ignore all the extra constraints I used (equal, coincident, the extra points, the hidden construction line), this is how I'd like it to look:
![image](https://user-images.githubusercontent.com/3342777/176531439-19d730a6-51eb-48b3-86b4-34c82bc82d2b.png)

**Comment by amrsoll** [comment]
To be fair, there is a simpler alternative:

You can add these constraints in either order
- Add parallel constraint
- Add distance constraint
  a. Select 1 end of the 1st line (point/vertice)
  b. Select the 2nd line
  c. Add the distance constraint
  > You have to select the point first, it doesn't work by selecting a line first

### Result

![image](https://user-images.githubusercontent.com/24782867/176538895-c2bc6d51-3bbf-4fd7-af9f-9fdbe110917d.png)

**Comment by TheMakerTales** [comment]
I 100% see the use case for this. I guess the question comes to should there be Line to line distance constraints?
It gets a little complicated when taking into consideration different line lengths to one another and if they are not parallel. Does other CAD software do line to line distancing? and if so how are they dealing with these problems?

**Reply by amrsoll** [reply]
I agree, the non-parallel line distance is not trivial.

The best I can come up with is 3 possibilities, all stemming from the fact that we can define Line - Point distance:
Assuming you select L1 and L2 (Lines), the disance constraint can be:
- Distance from L1 to centerpoint of L2
- Distance from L1 to furthest extremity of L2
- Distance from L1 to closest extremity of L2

Any of these possibilities are compatible with the distance between parallel lines (except if lines are infinite)
It would be very interesting to know how other software defines it.

**Reply by Intuos5** [reply]
Yes they are. Fusion 360 even does offset constraints. So if you use the offset tool, it creates a constraint that ties both curves together. It then adds a distance constraint, which drives the overall offset. For this to work, it always computes the second curve from the original curve.

**Comment by amrsoll** [comment]
TL;DR Makes me think we could introduce a new constraint: Size. That is effectively how the distance constraint works for 1 entity.

I am currently playing with the constraint system, including line distances. It seems like there can be 2 schools of thought for the "distance" constraint. Either you control the size of the element, or the distance of the element to other stuff.

The constraint if used either as a **size** or a **distance** constraint. Both methods can be generalised to 2 body systems, but only the size constraint makes sense in a single body problem.

Size constraint in an N body system: Constraints the size of N element relative to an other (like locking the display ratio of a picture when changing it's size). Constraining 1 element locks it's size (removes 1 degree of freedom). I could go on with other interpretations, like projections etc...

**Reply by Intuos5** [reply]
Fusion 360 has a user preference to 'size' the sketch with the first dimension. I believe this request came from Solidworks users back in the day.
